src/Servers/IIS/IIS/src/Core/IISHttpServer.cs

```diff
@@ -110,22 +113,11 @@ void RegisterCancelation()
                 return _shutdownSignal.Task;
             }
 
-            // First call back into native saying "DON'T SEND ME ANY MORE REQUESTS"
             _nativeApplication.StopIncomingRequests();
 
             try
             {
-                // Wait for active requests to drain
-                if (_outstandingRequests > 0)
-                {
-                    RegisterCancelation();
-                }
-                else
-                {
-                    // We have drained all requests. Block any callbacks into managed at this point.
-                    _nativeApplication.StopCallsIntoManaged();
-                    _shutdownSignal.TrySetResult(null);
-                }
+                RegisterCancelation();
```

Comment:
Do we need ` if (Interlocked.Exchange(ref _stopping, 1) == 1)` logic before that?

Reply:
We don't have test that uses cancellation token.

Reply:
> Do we need if (Interlocked.Exchange(ref _stopping, 1) == 1) logic before that?

We don't, but it avoids calling into native again and calling into native after Dispose.

> We don't have test that uses cancellation token.

I'll add one. We didn't have a test that called StopAsync at all, so I'll add a second test where we use the cancellation token.

> Store the registration and dispose it in OnDrainComplete

 Yeah that looks like a mistake.

Reply:
There are too many places we call:

```
server._nativeApplication.StopCallsIntoManaged();
server._shutdownSignal.TrySetResult(null);
```

We should unify all of them and simplify this entire thing.

The problem with calling this early is that we would be still left with hanging request handlers so nothing would really stop.

Reply:
We shouldn't fix everything in this PR but let's not multiply duplicated code and understand exactly what we are doing.

Reply:
> The problem with calling this early is that we would be still left with hanging request handlers so nothing would really stop.

Yeah you are definitely right. I think we should only call both in RequestDrainedHandler

Reply:
Let's chat tomorrow about this. I think I resolved issues here (don't feel great about tests though).